FlickrViewer/DetailCollectionViewCell.swift

```diff
@@ -19,63 +21,132 @@ class DetailCollectionViewCell: UICollectionViewCell, UIScrollViewDelegate {
     
     
     @IBAction func detailViewClosing(_ sender: UIButton) {
         detailDelegate?.close()
     }
-    var detailDelegate: DetailViewCellDelegate?
     
+    var detailDelegate: DetailViewCellDelegate?
     var definedSize: CGSize?
+    var isImageZoomed = false
+    var isTopViewHidden = false
     
     override func awakeFromNib() {
         super.awakeFromNib()
+        
+        //MARK - gradient top
+        let topGradientLayer = CAGradientLayer()
+        topGradientLayer.colors = [UIColor.black.cgColor, UIColor.clear.cgColor]
+        topGradientLayer.startPoint = CGPoint(x: 1, y: 0)
+        topGradientLayer.endPoint = CGPoint(x: 1, y: 1)
+        topGradientLayer.frame = topGradientView.bounds
+        topGradientView.layer.addSublayer(topGradientLayer)
+        
+        //MARK - gradient bottom
+        let bottomGradientLayer = CAGradientLayer()
+        bottomGradientLayer.colors = [UIColor.black.cgColor, UIColor.clear.cgColor]
+        bottomGradientLayer.startPoint = CGPoint(x: 1, y: 1)
+        bottomGradientLayer.endPoint = CGPoint(x: 1, y: 0)
+        bottomGradientLayer.frame = bottomGradientView.bounds
+        bottomGradientView.layer.addSublayer(bottomGradientLayer)
+        
+        scrollView.minimumZoomScale = 1.0
+        scrollView.maximumZoomScale = 3.0
+        
         let tapRecognizer = UITapGestureRecognizer(target: self, action: #selector(tap))
+        tapRecognizer.numberOfTapsRequired = 1
         self.addGestureRecognizer(tapRecognizer)
-        scrollView.minimumZoomScale = 1.0
-        scrollView.maximumZoomScale = 6.0
-        scrollView.delegate = self
+        
+        let doubleTapRecognizer = UITapGestureRecognizer(target: self, action: #selector(doubleTap))
+        doubleTapRecognizer.numberOfTapsRequired = 2
+        self.addGestureRecognizer(doubleTapRecognizer)
+        
+        tapRecognizer.require(toFail: doubleTapRecognizer)
+    }
+    
+   
+    
+    func scrollViewDidZoom(_ scrollView: UIScrollView) {
+        if scrollView.zoomScale > 1 {
+            isImageZoomed = true
+        } else {
+            isImageZoomed = false
+        }
     }
     
-    @objc private func tap(){
+    func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        
+        let leftMargin = (scrollView.bounds.size.width - scrollView.contentSize.width) * 0.5
+        let topMargin = (scrollView.bounds.size.height - scrollView.contentSize.height) * 0.5
+        scrollView.contentInset = UIEdgeInsets(top: max(0, topMargin), left: max(0, leftMargin), bottom: 0, right: 0)
+    }
+    
+    func viewForZooming(in scrollView: UIScrollView) -> UIView? {
+        return currentImage
+    }
+    
+    @objc private func tap() {
+        guard let detailDelegate = detailDelegate else {
+            return
+        }
         topView.isHidden = !topView.isHidden
+        detailDelegate.isTopViewHidden = !detailDelegate.isTopViewHidden
         bottomView.isHidden = !bottomView.isHidden
```

Comment:
bottomView.isHidden = !bottomView.isHidden 
change to 
bottomView.isHidden = detailDelegate.isTopViewHidden
dont use UI as data container